File: 프로그래머스/3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.py
def solution(s):
    n = len(s)
    
    dp = [[False] * n for _ in range(n)]

    for i in range(n):
        dp[i][i] = True

    longest_length = 1

    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            longest_length = 2

    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            if dp[i + 1][j - 1] and s[i] == s[j]:
                dp[i][j] = True
                longest_length = length

    return longest_length

File: 프로그래머스/3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

// 파라미터로 주어지는 문자열은 const로 주어집니다. 변경하려면 문자열을 복사해서 사용하세요.
int solution(const char* s) {
    int n = strlen(s);

    bool dp[n][n];
    
    int longestLength = 1;

    for (int i = 0; i < n; i++) {
        dp[i][i] = true;
    }
    
    for (int length = 2; length <= n; length++) {
        for (int i = 0; i <= n - length; i++) {
            int j = i + length - 1;

            dp[i][j] = (s[i] == s[j] && (length == 2 || dp[i + 1][j - 1]));
            if (dp[i][j]) {
                longestLength = length;
            }
        }
    }

    return longestLength;
}